pkg/controller/controller.go

```diff
@@ -210,64 +214,108 @@ func worker(queue workqueue.RateLimitingInterface, resourceType string, maxRetri
 // getServiceClassPlanAndServiceBroker is a sequence of operations that's done in couple of
 // places so this method fetches the Service Class, Service Plan and creates
 // a brokerClient to use for that method given an ServiceInstance.
-func (c *controller) getServiceClassPlanAndServiceBroker(instance *v1alpha1.ServiceInstance) (*v1alpha1.ServiceClass, *v1alpha1.ServicePlan, string, osb.Client, error) {
+func (c *controller) getServiceClassPlanAndServiceBroker(instance *v1alpha1.ServiceInstance, updateWithClone bool) (*v1alpha1.ServiceClass, *v1alpha1.ServicePlan, string, osb.Client, error) {
```

Comment:
Instead of having a flag here to decide if we should clone the object, I think we should make it required for the object passed into here to be a clone. We've reached a point in my opinion where the additional cloning is polluting our codebase. If we just clone after we've checked to make sure we will be in fact reconciling a resource, then we can clone once and not have to worry.

Regardless, though I do see that this fixes some bugs should we have not cloned yet before running this function, I would say that it is out of the scope of this PR and should be moved to a separate one.

Reply:
It seemed to me that there was a concerted effort to only do the resource cloning when absolutely necessary. There are some cases in the code path that calls this function where the clone is not needed. If we do not care as much about ensuring that we do not clone when not necessary, then I am more than happy to remove the updateWithClone parameter. It certainly makes the code cleaner.

Reply:
We shouldn't clone more than necessary, but the only reason we've up until now had a ton of littered clonings is because we've done a bad job of vetting when an update is necessary or not. The flow should be:

- Is there work to do?
     - If no, return.
     - If yes, clone the instance, and do the work.

The act of doing work on an instance implies that the status will be updated in some way (except for when we poll and don't get a description back, which I argue we should still update the status to capture the transition time). I have a yet-unpushed set of changes that do exactly that, but this change will cause a nasty rebase so I'd prefer holding off. But I'd remove this part anyway because it's simply out of scope of the main intent of your PR.

Reply:
I have removed the cloneToUpdate parameter from this function.